### PR TITLE
[SPARK-30352][SQL] DataSourceV2: Add CURRENT_CATALOG function

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

As we support multiple catalogs with DataSourceV2, we may need the `CURRENT_CATALOG` value expression from the SQL standard.

`CURRENT_CATALOG` is a general value specification in the SQL Standard, described as:

> The value specified by CURRENT_CATALOG is the character string that represents the current default catalog name.

### Why are the changes needed?
improve catalog v2 with ANSI SQL standard.

### Does this PR introduce any user-facing change?
yes, add a new function `current_catalog()` to point the current active catalog

### How was this patch tested?

add ut

### CODE REVIEW

**Comment by yaooqinn** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by yaooqinn** [comment]
retest this please,  and thanks @HyukjinKwon

**Comment by HyukjinKwon** [comment]
cc @cloud-fan FYI

**Comment by cloud-fan** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -130,6 +130,7 @@ abstract class Optimizer(catalogManager: CatalogManager)
       ReplaceExpressions,
       ComputeCurrentTime,
       GetCurrentDatabase(catalogManager),
+      GetCurrentCatalog(catalogManager),
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:133]: can we merge the above 2 rules into one?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 2]
```diff
@@ -116,6 +116,24 @@ case class CurrentDatabase() extends LeafExpression with Unevaluable {
...
+ * Returns the current catalog.
+ */
+@ExpressionDescription(
+  usage = "_FUNC_() - Returns the current catalog.",
+  examples = """
+    Examples:
+      > SELECT _FUNC_();
+       spark_catalog
+  """,
+  since = "3.0.0")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala:129]: it's too late for 3.0 now. let's change it to `3.1.0`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala [hunk 3]
```diff
@@ -91,15 +91,21 @@ object ComputeCurrentTime extends Rule[LogicalPlan] {
 }
 
 
-/** Replaces the expression of CurrentDatabase with the current database name. */
-case class GetCurrentDatabase(catalogManager: CatalogManager) extends Rule[LogicalPlan] {
+/**
+ * Replaces the expression of CurrentDatabase with the current database name.
+ * Replaces the expression of CurrentCatalog with the current catalog name.
+ */
+case class GetCurrentDatabaseOrCatalog(catalogManager: CatalogManager) extends Rule[LogicalPlan] {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala:98]: `GetCurrentDatebaseAndCatalog`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala [hunk 4]
```diff
@@ -91,15 +91,21 @@ object ComputeCurrentTime extends Rule[LogicalPlan] {
...
   def apply(plan: LogicalPlan): LogicalPlan = {
     import org.apache.spark.sql.connector.catalog.CatalogV2Implicits._
-    val currentNamespace = catalogManager.currentNamespace.quoted
 
     plan transformAllExpressions {
       case CurrentDatabase() =>
+        val currentNamespace = catalogManager.currentNamespace.quoted
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala:104]: it's better to get the value before `transformAllExpressions`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala [hunk 5]
```diff
@@ -91,15 +91,21 @@ object ComputeCurrentTime extends Rule[LogicalPlan] {
...
         Literal.create(currentNamespace, StringType)
+      case CurrentCatalog() =>
+        val currentCatalog = catalogManager.currentCatalog.name()
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala:107]: ditto

File: sql/core/src/test/resources/sql-tests/inputs/misc-functions.sql [hunk 6]
```diff
@@ -8,3 +8,6 @@ select typeof(cast(1.0 as float)), typeof(1.0D), typeof(1.2);
 select typeof(date '1986-05-23'),  typeof(timestamp '1986-05-23'), typeof(interval '23 days');
 select typeof(x'ABCD'), typeof('SPARK');
 select typeof(array(1, 2)), typeof(map(1, 2)), typeof(named_struct('a', 1, 'b', 'spark'));
+
+-- get current_catalog
+select current_catalog();
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/misc-functions.sql:13]: do we have a sql testing file for `current_database`?

Comment (yaooqinn) [reply sql/core/src/test/resources/sql-tests/inputs/misc-functions.sql:13]: checked and not found

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/inputs/misc-functions.sql:13]: how about we add a `current_database_catalog.sql` and test both of them there?

Comment (yaooqinn) [reply sql/core/src/test/resources/sql-tests/inputs/misc-functions.sql:13]: I am okay with it.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
sorry it conflicts, can you fix it?

**Comment by yaooqinn** [comment]
retest this please

**Comment by yaooqinn** [comment]
retest this please

**Comment by yaooqinn** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**